File: js/src/utils/fileFormats.ts
export const fileFormats: Record<string, string> = {
    // Documents
    pdf: "application/pdf",
    doc: "application/msword",
    docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    xls: "application/vnd.ms-excel",
    xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    ppt: "application/vnd.ms-powerpoint",
    pptx: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    csv: "text/csv",
    txt: "text/plain",
    rtf: "application/rtf",

    // Images
    png: "image/png",
    jpeg: "image/jpeg",
    jpg: "image/jpeg",
    gif: "image/gif",
    webp: "image/webp",
    svg: "image/svg+xml",
    bmp: "image/bmp",
    tiff: "image/tiff",
    ico: "image/x-icon",

    // Audio
    mp3: "audio/mpeg",
    wav: "audio/wav",
    ogg: "audio/ogg",
    aac: "audio/aac",
    wma: "audio/x-ms-wma",

    // Video
    mp4: "video/mp4",
    avi: "video/x-msvideo",
    mov: "video/quicktime",
    wmv: "video/x-ms-wmv",
    flv: "video/x-flv",
    webm: "video/webm",

    // Archives
    zip: "application/zip",
    rar: "application/vnd.rar",
    "7z": "application/x-7z-compressed",
    tar: "application/x-tar",
    gz: "application/gzip",

    // Code
    json: "application/json",
    xml: "application/xml",
    html: "text/html",
    css: "text/css",
    js: "application/javascript",
    ts: "application/typescript",
    py: "text/x-python",
    java: "text/x-java-source",
    cpp: "text/x-c++src",
    c: "text/x-csrc",

    // Other
    sql: "application/sql",
    yaml: "application/x-yaml",
    yml: "application/x-yaml",
    md: "text/markdown",
    log: "text/plain"
};

import { InvalidParameterError, EmptyCollectionError } from '../errors';

export class FileFormatManager {
    private static instance: FileFormatManager;
    private customFormats: Map<string, string> = new Map();

    private constructor() { }

    static getInstance(): FileFormatManager {
        if (!FileFormatManager.instance) {
            FileFormatManager.instance = new FileFormatManager();
        }
        return FileFormatManager.instance;
    }

    /**
     * Gets the MIME types for a list of formats
     */
    static getMimeTypes(formats: string[]): string[] {
        if (!formats || !Array.isArray(formats) || formats.length === 0) {
            throw new EmptyCollectionError('formats', 'At least one format must be provided');
        }

        const manager = FileFormatManager.getInstance();
        return formats.map(format => {
            const mimeType = manager.getMimeType(format);
            if (!mimeType) {
                throw new InvalidParameterError('format', format, `Unsupported file format: ${format}`);
            }
            return mimeType;
        });
    }

    /**
     * Gets the MIME type for a specific format
     */
    getMimeType(format: string): string | undefined {
        const normalizedFormat = format.toLowerCase().replace(/^\./, '');

        // Check custom formats first
        if (this.customFormats.has(normalizedFormat)) {
            return this.customFormats.get(normalizedFormat);
        }

        // Then the standard formats
        return fileFormats[normalizedFormat];
    }

    /**
     * Adds a custom format
     */
    addCustomFormat(extension: string, mimeType: string): void {
        const normalizedExtension = extension.toLowerCase().replace(/^\./, '');
        this.customFormats.set(normalizedExtension, mimeType);
    }

    /**
     * Removes a custom format
     */
    removeCustomFormat(extension: string): boolean {
        const normalizedExtension = extension.toLowerCase().replace(/^\./, '');
        return this.customFormats.delete(normalizedExtension);
    }

    /**
     * Checks whether a format is supported
     */
    isSupported(format: string): boolean {
        return this.getMimeType(format) !== undefined;
    }

    /**
     * Gets all supported formats
     */
    getSupportedFormats(): string[] {
        const standardFormats = Object.keys(fileFormats);
        const customFormats = Array.from(this.customFormats.keys());
        return [...standardFormats, ...customFormats];
    }

    /**
     * Gets formats by category
     */
    getFormatsByCategory(): Record<string, string[]> {
        return {
            documents: ['pdf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'csv', 'txt', 'rtf'],
            images: ['png', 'jpeg', 'jpg', 'gif', 'webp', 'svg', 'bmp', 'tiff', 'ico'],
            audio: ['mp3', 'wav', 'ogg', 'aac', 'wma'],
            video: ['mp4', 'avi', 'mov', 'wmv', 'flv', 'webm'],
            archives: ['zip', 'rar', '7z', 'tar', 'gz'],
            code: ['json', 'xml', 'html', 'css', 'js', 'ts', 'py', 'java', 'cpp', 'c'],
            other: ['sql', 'yaml', 'yml', 'md', 'log']
        };
    }

    /**
     * Validates a list of formats
     */
    validateFormats(formats: string[]): { valid: string[]; invalid: string[] } {
        const valid: string[] = [];
        const invalid: string[] = [];

        formats.forEach(format => {
            if (this.isSupported(format)) {
                valid.push(format);
            } else {
                invalid.push(format);
            }
        });

        return { valid, invalid };
    }

    /**
     * Gets the extension from a MIME type
     */
    getExtensionFromMimeType(mimeType: string): string | undefined {
        // Look in the standard formats
        for (const [ext, mime] of Object.entries(fileFormats)) {
            if (mime === mimeType) {
                return ext;
            }
        }

        // Look in the custom formats
        for (const [ext, mime] of this.customFormats.entries()) {
            if (mime === mimeType) {
                return ext;
            }
        }

        return undefined;
    }

    /**
     * Gets MIME types by category
     */
    getMimeTypesByCategory(): Record<string, string[]> {
        const categories = this.getFormatsByCategory();
        const result: Record<string, string[]> = {};

        for (const [category, formats] of Object.entries(categories)) {
            result[category] = formats
                .map(format => this.getMimeType(format))
                .filter((mime): mime is string => mime !== undefined);
        }

        return result;
    }
}

// Utility function for backward compatibility with legacy code
export function getMimeTypes(formats: string[]): string[] {
    return FileFormatManager.getMimeTypes(formats);
} 

File: py/yeriasdk/utils/file_formats.py
"""
File format management utilities
"""

from typing import Dict, List, Optional
from ..errors.exceptions import EmptyCollectionError, InvalidParameterError

# File format to MIME type mapping
FILE_FORMATS: Dict[str, str] = {
    # Documents
    "pdf": "application/pdf",
    "doc": "application/msword",
    "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "xls": "application/vnd.ms-excel",
    "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "ppt": "application/vnd.ms-powerpoint",
    "pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    "csv": "text/csv",
    "txt": "text/plain",
    "rtf": "application/rtf",
    # Images
    "png": "image/png",
    "jpeg": "image/jpeg",
    "jpg": "image/jpeg",
    "gif": "image/gif",
    "webp": "image/webp",
    "svg": "image/svg+xml",
    "bmp": "image/bmp",
    "tiff": "image/tiff",
    "ico": "image/x-icon",
    # Audio
    "mp3": "audio/mpeg",
    "wav": "audio/wav",
    "ogg": "audio/ogg",
    "aac": "audio/aac",
    "wma": "audio/x-ms-wma",
    # Video
    "mp4": "video/mp4",
    "avi": "video/x-msvideo",
    "mov": "video/quicktime",
    "wmv": "video/x-ms-wmv",
    "flv": "video/x-flv",
    "webm": "video/webm",
    # Archives
    "zip": "application/zip",
    "rar": "application/vnd.rar",
    "7z": "application/x-7z-compressed",
    "tar": "application/x-tar",
    "gz": "application/gzip",
    # Code
    "json": "application/json",
    "xml": "application/xml",
    "html": "text/html",
    "css": "text/css",
    "js": "application/javascript",
    "ts": "application/typescript",
    "py": "text/x-python",
    "java": "text/x-java-source",
    "cpp": "text/x-c++src",
    "c": "text/x-csrc",
    # Other
    "sql": "application/sql",
    "yaml": "application/x-yaml",
    "yml": "application/x-yaml",
    "md": "text/markdown",
    "log": "text/plain",
}


class FileFormatManager:
    """Singleton manager for file format to MIME type mapping"""

    _instance: Optional["FileFormatManager"] = None
    _custom_formats: Dict[str, str]

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._custom_formats = {}
        return cls._instance

    @classmethod
    def get_instance(cls) -> "FileFormatManager":
        """Get singleton instance"""
        return cls()

    @staticmethod
    def get_mime_types(formats: List[str]) -> List[str]:
        """Get MIME types for a list of formats"""
        if not formats or len(formats) == 0:
            raise EmptyCollectionError("formats", "At least one format must be provided")

        manager = FileFormatManager.get_instance()
        result = []
        for format_str in formats:
            mime_type = manager.get_mime_type(format_str)
            if not mime_type:
                raise InvalidParameterError(
                    "format", format_str, f"Unsupported file format: {format_str}"
                )
            result.append(mime_type)
        return result

    def get_mime_type(self, format_str: str) -> Optional[str]:
        """Get MIME type for a specific format"""
        normalized = format_str.lower().lstrip(".")

        # Check custom formats first
        if normalized in self._custom_formats:
            return self._custom_formats[normalized]

        # Then standard formats
        return FILE_FORMATS.get(normalized)

    def add_custom_format(self, extension: str, mime_type: str) -> None:
        """Add a custom format"""
        normalized = extension.lower().lstrip(".")
        self._custom_formats[normalized] = mime_type

    def remove_custom_format(self, extension: str) -> bool:
        """Remove a custom format"""
        normalized = extension.lower().lstrip(".")
        if normalized in self._custom_formats:
            del self._custom_formats[normalized]
            return True
        return False

    def is_supported(self, format_str: str) -> bool:
        """Check if a format is supported"""
        return self.get_mime_type(format_str) is not None

    def get_supported_formats(self) -> List[str]:
        """Get all supported formats"""
        standard = list(FILE_FORMATS.keys())
        custom = list(self._custom_formats.keys())
        return standard + custom

    def get_formats_by_category(self) -> Dict[str, List[str]]:
        """Get formats grouped by category"""
        return {
            "documents": [
                "pdf",
                "doc",
                "docx",
                "xls",
                "xlsx",
                "ppt",
                "pptx",
                "csv",
                "txt",
                "rtf",
            ],
            "images": ["png", "jpeg", "jpg", "gif", "webp", "svg", "bmp", "tiff", "ico"],
            "audio": ["mp3", "wav", "ogg", "aac", "wma"],
            "video": ["mp4", "avi", "mov", "wmv", "flv", "webm"],
            "archives": ["zip", "rar", "7z", "tar", "gz"],
            "code": ["json", "xml", "html", "css", "js", "ts", "py", "java", "cpp", "c"],
            "other": ["sql", "yaml", "yml", "md", "log"],
        }

    def validate_formats(self, formats: List[str]) -> Dict[str, List[str]]:
        """Validate a list of formats"""
        valid: List[str] = []
        invalid: List[str] = []

        for format_str in formats:
            if self.is_supported(format_str):
                valid.append(format_str)
            else:
                invalid.append(format_str)

        return {"valid": valid, "invalid": invalid}

    def get_extension_from_mime_type(self, mime_type: str) -> Optional[str]:
        """Get extension from MIME type"""
        # Check standard formats
        for ext, mime in FILE_FORMATS.items():
            if mime == mime_type:
                return ext

        # Check custom formats
        for ext, mime in self._custom_formats.items():
            if mime == mime_type:
                return ext

        return None

    def get_mime_types_by_category(self) -> Dict[str, List[str]]:
        """Get MIME types grouped by category"""
        categories = self.get_formats_by_category()
        result: Dict[str, List[str]] = {}

        for category, formats in categories.items():
            mime_types = []
            for format_str in formats:
                mime_type = self.get_mime_type(format_str)
                if mime_type:
                    mime_types.append(mime_type)
            result[category] = mime_types

        return result

